web/src/main/java/org/cbioportal/web/util/ClinicalDataBinUtil.java

```diff
@@ -0,0 +1,351 @@
+        List<String> unfilteredUniquePatientKeys = new ArrayList<>();
+        List<String> unfilteredSampleAttributeIds = new ArrayList<>();
+        List<String> unfilteredPatientAttributeIds = new ArrayList<>();
+        // patient attributes which are also sample attributes in other studies
+        List<String> unfilteredConflictingPatientAttributeIds = new ArrayList<>();
+
+        populateIdLists(
+            // input
+            unfilteredSampleIdentifiers,
+            attributeIds,
+
+            // output
+            unfilteredStudyIds,
+            unfilteredSampleIds,
+            unfilteredPatientIds,
+            studyIdsOfUnfilteredPatients,
+            unfilteredUniqueSampleKeys,
+            unfilteredUniquePatientKeys,
+            unfilteredSampleAttributeIds,
+            unfilteredPatientAttributeIds,
+            unfilteredConflictingPatientAttributeIds
+        );
+
+        Map<String, ClinicalDataType> attributeDatatypeMap = constructAttributeDataMap(
+            unfilteredSampleAttributeIds,
+            unfilteredPatientAttributeIds,
+            unfilteredConflictingPatientAttributeIds
+        );
+
+        List<ClinicalData> unfilteredClinicalDataForSamples = clinicalDataFetcher.fetchClinicalDataForSamples(
+            unfilteredStudyIds,
+            unfilteredSampleIds,
+            new ArrayList<>(unfilteredSampleAttributeIds)
+        );
+
+        List<ClinicalData> unfilteredClinicalDataForPatients = clinicalDataFetcher.fetchClinicalDataForPatients(
+            studyIdsOfUnfilteredPatients,
+            unfilteredPatientIds,
+            new ArrayList<>(unfilteredPatientAttributeIds)
+        );
+
+        List<ClinicalData> unfilteredClinicalDataForConflictingPatientAttributes = clinicalDataFetcher.fetchClinicalDataForConflictingPatientAttributes(
+            studyIdsOfUnfilteredPatients,
+            unfilteredPatientIds,
+            new ArrayList<>(unfilteredConflictingPatientAttributeIds)
+        );
+
+        List<ClinicalData> unfilteredClinicalData = Stream.of(
+            unfilteredClinicalDataForSamples,
+            unfilteredClinicalDataForPatients,
+            unfilteredClinicalDataForConflictingPatientAttributes
+        ).flatMap(Collection::stream).collect(Collectors.toList());
+
+        // if filters are practically the same no need to re-apply them
+        List<SampleIdentifier> filteredSampleIdentifiers =
+            studyViewFilterUtil.shouldSkipFilterForClinicalDataBins(studyViewFilter) ?
+                unfilteredSampleIdentifiers : studyViewFilterApplier.apply(studyViewFilter);
+
+        List<String> filteredUniqueSampleKeys;
+        List<String> filteredUniquePatientKeys;
+        List<ClinicalData> filteredClinicalData;
+
+        // if filtered and unfiltered samples are exactly the same, no need to fetch clinical data again
+        if (filteredSampleIdentifiers.equals(unfilteredSampleIdentifiers)) {
```

Comment:
may be replace this by `studyViewFilterUtil.shouldSkipFilterForClinicalDataBins(studyViewFilter)` and calculate `filteredSampleIdentifiers` inside else block

Reply:
In theory, `filteredSampleIdentifiers` may still be equal to `unfilteredSampleIdentifiers` even after applying the filter (in case provided filter does not filter anything out after we apply it). So ideally we should always check for equality for both lists. I am not sure if this ever happens in practice though.